src/authentication.ts

```diff
@@ -70,16 +75,35 @@ export default class GitpodAuthenticationProvider extends Disposable implements
 	async getSessions(scopes?: string[]): Promise<vscode.AuthenticationSession[]> {
 		// For the Gitpod scope list, order doesn't matter so we immediately sort the scopes
 		const sortedScopes = scopes?.sort() || [];
-		this._logger.info(`Getting sessions for ${sortedScopes.length ? sortedScopes.join(',') : 'all scopes'}...`);
+		const validScopes = await this.fetchValidScopes();
+		const sortedFilteredScopes = sortedScopes.filter(s => !validScopes || validScopes.includes(s));
+		this._logger.info(`Getting sessions for ${sortedScopes.length ? sortedScopes.join(',') : 'all scopes'}${sortedScopes.length !== sortedFilteredScopes.length ? `, but valid scopes are ${sortedFilteredScopes.join(',')}` : ''}...`);
+		if (sortedScopes.length !== sortedFilteredScopes.length) {
+			this._logger.warn(`But valid scopes are ${sortedFilteredScopes.join(',')}, returning session with only valid scopes...`);
+		}
 		const sessions = await this._sessionsPromise;
-		const finalSessions = sortedScopes.length
-			? sessions.filter(session => arrayEquals([...session.scopes].sort(), sortedScopes))
+		const finalSessions = sortedFilteredScopes.length
+			? sessions.filter(session => arrayEquals([...session.scopes].sort(), sortedFilteredScopes))
 			: sessions;
 
-		this._logger.info(`Got ${finalSessions.length} sessions for ${sortedScopes?.join(',') ?? 'all scopes'}...`);
+		this._logger.info(`Got ${finalSessions.length} sessions for ${sortedFilteredScopes?.join(',') ?? 'all scopes'}...`);
 		return finalSessions;
 	}
 
+	private async fetchValidScopes(): Promise<string[] | undefined> {
```

Comment:
Why not using the same function like `isOauthInspectSupported`? It's duplicate here.

And we can public a method here like `supportScope(...scopes)`, and verify everything here instead of call `isOauthInspectSupported` / `isFeatureSupported` in where we use it to make it easier to use

Reply:
> Why not using the same function like isOauthInspectSupported? It's duplicate here.

`isOauthInspectSupported` is a hack just for now as we need to be backwards compatible, it shouldn't be called at all in remoteConnector.ts

> And we can public a method here like supportScope(...scopes), and verify everything here instead of call isOauthInspectSupported / isFeatureSupported in where we use it to make it easier to use

Not sure what you mean here, `GitpodAuthenticationProvider` only implements `vscode.AuthenticationProvider` and it will be called by vscode internally